CronJob: Respect ControllerRef

enisoc opened the pull request:
**What this PR does / why we need it**:

This is part of the completion of the [ControllerRef](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md) proposal. It brings CronJob into compliance with ControllerRef. See the individual commit messages for details.

**Which issue this PR fixes**:

This ensures that other controllers do not fight over control of objects that a CronJob owns.

**Special notes for your reviewer**:

**Release note**:

```release-note
CronJob controller now respects ControllerRef to avoid fighting with other controllers.
```
cc @erictune @kubernetes/sig-apps-pr-reviews

Commit:
CronJob: Don't launch if DeletionTimestamp is set.

Commit:
CronJob: Add ControllerRef on all created Jobs.

Commit:
CronJob: Set DefaultGarbageCollectionPolicy to OrphanDependents.

Now that CronJob adds ControllerRef to Jobs it creates,
we need to set this default so legacy behavior is maintained.

Commit:
CronJob: Apply missing ControllerRefs to Jobs created by a CronJob.

This should only happen if the Jobs were created by an older version
of the CronJob controller, since from now on we add ControllerRef upon
creation.

CronJob doesn't do actual adoption because it doesn't use label
selectors to find its Jobs. However, we should apply ControllerRef
for potential server-side cascading deletion, and to advise other
controllers we own these objects.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42177)
<!-- Reviewable:end -->

soltysh reviewed (approved):
LGTM

soltysh commented:
/approve

soltysh commented:
/lgtm

kow3ns commented:
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---

*[pkg/controller/cronjob/cronjob_controller.go, line 58 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42177#-Ke0ct9rJOGoAhuktRoo:-Ke0ct9rJOGoAhuktRop:b1nbxya) ([raw file](https://github.com/kubernetes/kubernetes/blob/d471655d43517f83da7877867bb2ad2355fb2f98/pkg/controller/cronjob/cronjob_controller.go#L58)):*
> ```Go
> 
> // ControllerKind contains the schema.GroupVersionKind for this controller type.
> var ControllerKind = batch.SchemeGroupVersion.WithKind("CronJob")
> ```

Probably should not be exported from package

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42177)*
<!-- Sent from Reviewable.io -->

calebamiles commented:
@k8s-bot kops aws e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot gce etcd3 e2e test this

Commit:
CronJob: Always set BlockOwnerDeletion in ControllerRef.

Commit:
CronJob: Use synchronous deletion for CronJobs in e2e.

This is needed now that the default is OrphanDependents.

enisoc commented:
@k8s-bot node e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot cvm gce e2e test this

Commit:
CronJob: Check ControllerRef Name and UID in unit test.

enisoc commented:
@soltysh Please take a look. I believe all comments are addressed.

soltysh commented:
/lgtm

soltysh commented:
@ethernetdan @calebamiles is this ok to merge for 1.6 release?

janetkuo commented on pkg/controller/cronjob/utils.go:
```diff
@@ -266,3 +282,35 @@ func (o byJobStartTime) Less(i, j int) bool {
+			continue
+		}
+		job.OwnerReferences = append(job.OwnerReferences, sjControllerRef)
+		updatedJob, err := jc.UpdateJob(job.Namespace, job)
```
Use `Patch` instead?

enisoc replied:
What is the advantage of Patch? I used to think it would retry ResourceVersion conflicts for you on the server-side, but I recently found out it doesn't. Given that, I prefer the Update() pattern in code (as opposed to CLI) because the changes are shown semantically (obj.field = value), rather than as a JSON string. We could add a Retry() wrapper here, but as mentioned in the comment, it isn't really necessary since we will sync all CronJobs again in 10 seconds, and this only ever matters the first time you upgrade anyway (at steady-state, CronJob will add ControllerRef to every Job on creation).

kargakis replied:
Are we set on syncing CronJobs every 10s forever?

enisoc replied:
The [CronJob Manifesto](https://github.com/kubernetes/kubernetes/blob/cf1160702b4e0d848c0ea5cdfe0dc0a0973765dd/pkg/controller/cronjob/cronjob_controller.go#L20) states this as a goal, but I admit that may change in the future. Should I add a comment to draw attention to `adoptJobs()` if someone changes the [periodic sync behavior](https://github.com/kubernetes/kubernetes/blob/cf1160702b4e0d848c0ea5cdfe0dc0a0973765dd/pkg/controller/cronjob/cronjob_controller.go#L96)?

janetkuo replied:
> What is the advantage of Patch? I used to think it would retry ResourceVersion conflicts for you on the server-side, but I recently found out it doesn't. 

This is surprising. I've been told that Patch is the best practice and apiserver will retry for you. Is it not the case? @caesarxuchao @lavalamp

deads2k replied:
> This is surprising. I've been told that Patch is the best practice and apiserver will retry for you. Is it not the case? @caesarxuchao @lavalamp

It does retry for you on non-conflicting patches and that non-conflicting retry makes sure you don't accidentally clobber a change from someone else.

enisoc replied:
Here are the threads where I read this:

https://github.com/kubernetes/kubernetes/pull/42458#issuecomment-283876410
https://github.com/kubernetes/kubernetes/pull/42447#issuecomment-283877125

From the test failure linked in the second comment:

```
Operation cannot be fulfilled on replicationcontrollers "slow-terminating-unready-pod": there is a meaningful conflict (firstResourceVersion: "8517", currentResourceVersion: "8519"):
 diff1={"metadata":{"resourceVersion":"8519"},"spec":{"replicas":0},"status":{"conditions":null,"fullyLabeledReplicas":null,"replicas":0}}
, diff2={"spec":{"replicas":0}}
```

janetkuo replied:
This change already LGTM, we can discuss this further somewhere else

janetkuo replied:
Now that we have more time, shall we switch to `patch`?

enisoc replied:
OK I will work on switching to patch.

enisoc replied:
Done.

kow3ns commented:
/lgtm

lavalamp commented on pkg/registry/batch/cronjob/strategy_test.go:
```diff
@@ -92,6 +93,13 @@ func TestCronJobStrategy(t *testing.T) {
+
+	// Make sure we correctly implement the interface.
+	// Otherwise a typo could silently change the default.
+	var gcds rest.GarbageCollectionDeleteStrategy = Strategy
```
This is best done at compile time by declaring

```go
var _ rest.GarbageCollectionDeleteStrategy = Strategy
```

in the main package.

Feel free to change in a followup, I won't block on it.

enisoc replied:
Ok I've added it to my list and will change it across all controllers in a followup PR.

lavalamp commented:
/approve

(registry changes)

kargakis commented:
@k8s-bot cvm gce e2e test this

kargakis commented on pkg/controller/cronjob/utils.go:
```diff
@@ -266,3 +282,35 @@ func (o byJobStartTime) Less(i, j int) bool {
+		job.OwnerReferences = append(job.OwnerReferences, sjControllerRef)
+		updatedJob, err := jc.UpdateJob(job.Namespace, job)
+		if err != nil {
+			// If there's a ResourceVersion or other error, don't bother retrying.
```
It is unfortunate though that for a 409, that is probably not related to the field you are updating, you'll have to wait until the next sync interval (10s).

kargakis commented:
Do we have a separate issue for deprecating and removing created-by annotations in favor of owner refs?

fejta commented:
/assign @lavalamp 

Do you remember why you added the do-not-merge label?

enisoc commented:
I've removed LGTM because I have some changes planned to bring this up to date with the other ControllerRef PRs that moved forward during the 1.6 cycle. I'll request re-review once I've pushed those changes.

Commit:
CronJob: List children before parents.

This prevents a race with the GC while it orphans dependents.

Commit:
CronJob: Add e2e test for adoption.

Currently, an e2e test is the only way to ensure we have the proper RBAC
permissions to adopt Jobs.

janetkuo commented:
> /assign @lavalamp
> Do you remember why you added the do-not-merge label?

See https://github.com/kubernetes/kubernetes/pull/42173#issuecomment-285140858 basically because it was code-freeze then

kargakis commented:
> Do we have a separate issue for deprecating and removing created-by annotations in favor of owner refs?

https://github.com/kubernetes/kubernetes/issues/44407

enisoc commented:
This is ready for final re-review.

I've added 2 commits to fix the equivalent of https://github.com/kubernetes/kubernetes/pull/42938 and https://github.com/kubernetes/kubernetes/pull/43034. I made no changes to the previous commits other than rebasing on master.

kargakis commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -101,14 +104,10 @@ func (jm *CronJobController) Run(stopCh <-chan struct{}) {
+	// Note that this only works because we are NOT using any caches here.
 	jl, err := jm.kubeClient.BatchV1().Jobs(metav1.NamespaceAll).List(metav1.ListOptions{})
 	if err != nil {
 		glog.Errorf("Error listing jobs")
```
The error is not printed here

kargakis commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -117,6 +116,14 @@ func (jm *CronJobController) syncAll() {
 
+	sjl, err := jm.kubeClient.BatchV2alpha1().CronJobs(metav1.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		glog.Errorf("Error listing cronjobs: %v", err)
```
Can you change glog.Errorf calls to utilruntime.HandleError (the latter is ratelimited among other things)

kargakis commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -237,6 +244,18 @@ func syncOne(sj *batchv2alpha1.CronJob, js []batchv1.Job, now time.Time, jc jobC
 
+	if sj.DeletionTimestamp != nil {
+		// The CronJob is being deleted.
+		// Don't do anything other than updating status.
```
There is not status update.

kargakis replied:
Oh well it seems it has already happened above

kargakis commented on test/e2e/cronjob.go:
```diff
@@ -274,6 +275,53 @@ var _ = framework.KubeDescribe("CronJob", func() {
+
+	// Adopt Jobs it owns that don't have ControllerRef yet.
+	// That is, the Jobs were created by a pre-v1.6.0 master.
+	It("should adopt Jobs it owns that don't have ControllerRef yet", func() {
```
This could also be an integration test, no?

enisoc replied:
This was mainly added because we've run into problems with RBAC before. Do integration tests cover that?

kargakis replied:
I would expect so, @deads2k ?

janetkuo commented on test/e2e/generated_clientset.go:
```diff
@@ -324,7 +324,8 @@ var _ = framework.KubeDescribe("Generated release_1_5 clientset", func() {
 		By("deleting the cronJob")
-		if err := cronJobClient.Delete(cronJob.Name, nil); err != nil {
+		// Use OrphanDependents=false so the delete is synchronous.
+		if err := cronJobClient.Delete(cronJob.Name, &metav1.DeleteOptions{OrphanDependents: new(bool)}); err != nil {
```
`OrphanDependents` is being deprecated in 1.7. Should we switch to use `deleteOptions.propagationPolicy` instead?

enisoc replied:
Done.

janetkuo commented on pkg/controller/cronjob/cronjob_controller.go:
```diff
@@ -105,21 +105,25 @@ func (jm *CronJobController) Run(stopCh <-chan struct{}) {
+	// List children (Jobs) before parents (CronJob).
+	// This guarantees that if we see any Job that got orphaned by the GC orphan finalizer,
+	// we must also see that the parent CronJob has non-nil DeletionTimestamp (see #42639).
+	// Note that this only works because we are NOT using any caches here.
```
@soltysh we'll need to take care of this when we change cronjob controller to use informer

soltysh replied:
Yes, we've talked about it during KubeCon. If I have a reasonable solution in place in openshift I'll make sure to fix cronjob controller.

janetkuo commented:
Do we need to test that SJ won't adopt Jobs it doesn't own?

enisoc commented:
> Do we need to test that SJ won't adopt Jobs it doesn't own?

This is unit tested via `TestGroupJobsByParent()`. Jobs that the CronJob doesn't own should not even be passed into `adoptJobs()`.

The e2e test is only to ensure surprises like RBAC don't prevent adopt calls from working. In this case, the correct action is to do nothing, so we don't need an e2e test to make sure other components don't interfere.

There is no "release/orphan" case since CronJob does not use a label selector to find its children.

Commit:
CronJob: Use PATCH to adopt Jobs.

janetkuo reviewed (approved):
LGTM

soltysh commented:
> Does this mean that whenever I delete a cronjob from a k8s cluster, all its jobs, and associated pods would get deleted too?

That depends how you remove the object, it is still possible to remove an object and orphan its dependents.

davidghiurco commented:
I delete the object using helm.
e.g.
`
helm delete --purge <helm_release_name>
`

soltysh commented:
Can you provide me the actual requests `helm delete` sends over to the server, I'm not familiar with helm.
